### PR TITLE
Fix array variable relaxation in Picard iteration

### LINKED ISSUE #17659: AllLocalDofIndicesThread does not work with ArrayVariables
## Bug Description

Using relaxation option in MultiApp for an array variable leads to error that variable is not known.
The reason is that `AllLocalDofIndicesThread` does not handle array variables correctly. 

## Steps to Reproduce

As above.

## Impact

Not certain, somewhere in between small annoyance and detrimental bug.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close #17659.

### CODE REVIEW

**Comment by YaqiWang** [comment]
This will replace #17660.

**Comment by YaqiWang** [comment]
This is ready for review. The failure in controlled app appears irrelevant.

**Review by fdkong — COMMENTED** [review]
A question

File: framework/src/systems/SystemBase.C [hunk 1]
```diff
@@ -251,7 +239,8 @@ SystemBase::zeroVariables(std::vector<std::string> & vars_to_be_zeroed)
   {
     NumericVector<Number> & solution = this->solution();
 
-    AllLocalDofIndicesThread aldit(system(), vars_to_be_zeroed, true);
+    AllLocalDofIndicesThread aldit(
+        dynamic_cast<FEProblemBase &>(_subproblem), vars_to_be_zeroed, true);
```
[the remark this answers is not in the document]
Comment (YaqiWang) [reply framework/src/systems/SystemBase.C:243]: ok

**Review by fdkong — APPROVED** [review]

**Comment by loganharbour** [comment]
This broke next: https://civet.inl.gov/event/67146/

**Comment by fdkong** [comment]
@YaqiWang Could you take a look into the next?

**Comment by YaqiWang** [comment]
Looking.

**Comment by YaqiWang** [comment]
That test seems revealing an existing valgrind issue in `Coupleable`: https://github.com/idaholab/moose/blob/next/framework/src/interfaces/Coupleable.C#L350. Do you think changing that line to the line like here (https://github.com/idaholab/moose/blob/next/framework/src/interfaces/Coupleable.C#L318) can fix the problem?

**Comment by GiudGiud** [comment]
I'm adding an assert in #17479, following the changes in this PR, to check for the existence of _sys when calling AllLocalDofIndicesThread() because this is no longer guaranteed when using the problem. _sys is only found when variables are specified, but sometimes there are none.

**Comment by YaqiWang** [comment]
Oh, passing none means to get all local dofs? You may want to just use the nonlinear system from FEProblemBase in that case.

**Comment by GiudGiud** [comment]
Passing no variables to relax. 
Yes that would be an option. An assert is fine for the next one running into this

**Pull request merged**